FeedStoreChallenge/CoreData/CoreDataFeedStore.swift

```diff
@@ -29,14 +29,59 @@ public final class CoreDataFeedStore: FeedStore {
 	}
 
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				if let managedCache = try ManagedCache.find(in: context) {
+					completion(
+						.found(
+							feed: managedCache.localFeed,
+							timestamp: managedCache.timestamp
+						)
+					)
+				} else {
+					completion(.empty)
+				}
+
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
 
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				let managedCache = try ManagedCache.newUniqueInstance(in: context)
+				managedCache.timestamp = timestamp
+				managedCache.feed = ManagedCache.images(
+					from: feed,
+					in: context
+				)
+				try context.save()
+				completion(nil)
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
 	}
 
 	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				try ManagedCache.find(in: context)
+					.map(context.delete(_:))
+					.map(context.save)
+				completion(nil)
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
+	}
+
+	private func perform(_ action: @escaping (NSManagedObjectContext) -> Void) {
+		let context = self.context
+		context.perform { action(context) }
```

Comment:
During the course this way of having a local reference to the `context` to avoid capturing `self` inside of the `context.perform` block is taught.
The same can be achieved using a capture list.
It saves one line and one variable.

No need to make this amendment, this is just for your interest.

```suggestion
		context.perform { [context] in  action(context) }
```